### PR TITLE
Check MF_DROPPED for all ammo and weapons.

### PR DESCRIPTION
**Pull request by NeuralStunner**
Allows DeHackEd mods to customize item drops better.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Thank you for the effort, but I am still not sure what this pretty invasive change is going to achieve.

**Comment by NeuralStunner** [comment]
It's less useful with the limited DOS DEH space, but for BEX mods one can, say, make arachnotrons drop partial plasma cells. Making a new thing with the DROPPED flag pre-applied ensures it won't be a full pickup (making things too easy).

I think I'm going to approach this differently: make the num argument for P_GiveAmmo into a float instead, then dropped large pickups can pass 2.5 instead.

**Comment by fabiangreffrath** [comment]
Sorry, but I have to disregard this patch (at least for now) for two reasons:

1. I find this wrong in a semantic way. The `MF_DROPPED` flag isn't a property for a thing class, it is a property for individual things. Either a thing is dropped or it is not. Are there any other source ports which support this questionable feature you propose here?
2. When you said you would make it less invasive, you actually made it worse. Sorry, but passing a `float` type argument to a function like `P_GiveAmmo()` is an absolute no-go.

**Comment by NeuralStunner** [comment]
1. Mobj flags are always initialized from the thing table. Everything is exposed through DEH (including internal ones like SKULLFLY). Some old mods use the player color flags in particular to recolor objects such as player sprites or barrels.
ZDoom and EDGE both evaluate the DROPPED flag for any kind of weapon or ammo drop. This is important for their native scripting systems, you can assign drops to new/replaced monsters and they'll "just work". Eternity does not seem to apply this flag on map pickups, not being familiar with EDF I'm unsure how custom half-ammo drops are made to function.
2. I initially tried making "dropped" an optional argument to P_GiveAmmo, but had no success. Since it's never called outside p_inter, I let it go at the time. Float was the only solution I saw that would allow outside calls to this function to expect a numeric argument and nothing extra.
Since the large ammo drops give an odd number of "clips" there isn't a great way to pass that detail to P_GiveAmmo as-is.
My last resort idea is to have it interpret any negative value as "give half the absolute value". It doesn't seem any less weird than 0 as "give half of 1".

**Comment by fabiangreffrath** [comment]
Have you made sure that this gives the *exact* same amount of ammo for any picked up weapon or ammo as Vanilla Doom if no DEH patch is loaded?
If you want to reduce the amount of ammo given on pick-up, why don't you just use DEH to change the "Per ammo" property?

Apart from these two questions, let me have a look at this. Maybe I can find an approach that pleases both of us. :wink:

File: src/doom/p_inter.c [hunk 1]
```diff
@@ -79,10 +80,12 @@ P_GiveAmmo
     if ( player->ammo[ammo] == player->maxammo[ammo]  )
 	return false;
 		
-    if (num)
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: I think the `if (num)` condition may remain, even if it won't be triggered anymore during usual gameplay (I am curently looking for this patch to become least-invasive).

Also, could you probably move the `if (dropped)` condition behind the `gameskill` check?

File: src/doom/p_inter.c [hunk 2]
```diff
@@ -207,10 +210,7 @@ P_GiveWeapon
     {
 	// give one clip with a dropped weapon,
 	// two clips with a found weapon
-	if (dropped)
-	    gaveammo = P_GiveAmmo (player, weaponinfo[weapon].ammo, 1);
-	else
-	    gaveammo = P_GiveAmmo (player, weaponinfo[weapon].ammo, 2);
+		gaveammo = P_GiveAmmo (player, weaponinfo[weapon].ammo, 2, dropped);
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: You could simply disable the `if (dropped)` code path by appending `&& false` to the condition.

File: src/doom/p_inter.c [hunk 3]
```diff
@@ -544,57 +544,57 @@ P_TouchSpecialThing
...
-	{
-	    if (!P_GiveAmmo (player,am_clip,0))
-		return;
-	}
-	else
-	{
-	    if (!P_GiveAmmo (player,am_clip,1))
-		return;
-	}
+	if (!P_GiveAmmo (player,am_clip, 1,
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: Please comment out this block with `/* ... */` instead of removing it.

File: src/doom/p_inter.c (hunk 1 above)
Comment (NeuralStunner) [reply src/doom/p_inter.c]: The last commit restores if (num) and wraps the new logic in it.

A negative 'num' is interpreted as "half the value of abs(num) clips". I don't *think* there's a reason to use this function for taking clips worth of ammo.

Still, if the 'dropped' parameter makes more sense, I can go back to that.

File: src/doom/p_inter.c (hunk 2 above)
Comment (NeuralStunner) [reply src/doom/p_inter.c]: The block is restored in the last commit, but I'll keep this in mind if you decide to have 'dropped' back on P_GiveAmmo.

File: src/doom/p_inter.c (hunk 3 above)
Comment (NeuralStunner) [reply src/doom/p_inter.c]: Again, back to using this for the time being.

File: src/doom/p_inter.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/doom/p_inter.c]: > Still, if the 'dropped' parameter makes more sense, I can go back to that.

Yes, please do so. Sorry for being so unsteady, but you changed the code faster than I could properly review it.

File: src/doom/p_inter.c [hunk 4]
```diff
@@ -59,14 +59,16 @@ int	clipammo[NUMAMMO] = {10, 4, 20, 1};
 // P_GiveAmmo
 // Num is the number of clip loads,
 // not the individual count (0= 1/2 clip).
+// [NS] Dropped ammo/weapons give half as much.
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: This comment should rather go right next to the `dropped` parameter in the function definiton. It gets lost here.

File: src/doom/p_inter.c [hunk 5]
```diff
@@ -207,14 +218,17 @@ P_GiveWeapon
...
+	/*
 	if (dropped)
 	    gaveammo = P_GiveAmmo (player, weaponinfo[weapon].ammo, 1);
 	else
 	    gaveammo = P_GiveAmmo (player, weaponinfo[weapon].ammo, 2);
+	*/
     }
     else
 	gaveammo = false;
-
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: No white space changes, please.

File: src/doom/p_inter.c [hunk 6]
```diff
@@ -352,6 +366,7 @@ P_TouchSpecialThing
     int		i;
     fixed_t	delta;
     int		sound;
+    boolean dropped;
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: `const boolean dropped = ((special->flags & MF_DROPPED) != 0);`

File: src/doom/p_inter.c [hunk 7]
```diff
@@ -607,58 +628,56 @@ P_TouchSpecialThing
...
       case SPR_BFUG:
-	if (!P_GiveWeapon (player, wp_bfg, false) )
+	if (!P_GiveWeapon (player, wp_bfg, dropped) )
 	    return;
 	player->message = DEH_String(GOTBFG9000);
 	sound = sfx_wpnup;	
 	break;
 	
       case SPR_MGUN:
-        if (!P_GiveWeapon(player, wp_chaingun,
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: Vanilla code must remain intact, please! This also applies to shotgun and supershotgun.

File: src/doom/p_inter.c [hunk 8]
```diff
@@ -93,6 +95,15 @@ P_GiveAmmo
     }
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: Could you move this between the two "blank" lines before?

File: src/doom/p_inter.c [hunk 9]
```diff
@@ -554,47 +572,50 @@ P_TouchSpecialThing
 	    if (!P_GiveAmmo (player,am_clip,1))
 		return;
 	}
+	*/
+	if (!P_GiveAmmo (player,am_clip,1,dropped))
```
Comment (fabiangreffrath) [inline src/doom/p_inter.c]: Please keep indentation as it was in the commented out code.

**Comment by fabiangreffrath** [comment]
Thank you for your patience!

**Comment by NeuralStunner** [comment]
Same!

**Pull request merged**